### PR TITLE
Fixed match for resources with special plural names

### PR DESCRIPTION
**Pull request by lukas-bednar**
* track unbound methods
* added tests module

### CODE REVIEW

**Comment by avi3tal** [comment]
Di you run tests on Foreman 1.5 1.6 and 1.7 schemas?

**Comment by lukas-bednar** [comment]
Yes I did, and also on that yours one, where I met some troubles I described to you above.
You can cherry-pick all my open pull requests in order to get all fixes I made.
You can also run the test which tests all json files stored under foreman/definitions directory.
The test just create instance of Foreman class and initiate all Resources with relevant methods.
I have to find the way how to verify sane status of each method which was generated as api entry point.

```
py.test -s tests/
```

**Comment by avi3tal** [comment]
Sorry for the delay,
I was using you fix and seams like it is working just great except i got the following errors while initiating the Foreman instance:

ERROR:root:There is API entry-point which doesn't match expected constraints. Got resource 'operatingsystem' expected reference 'operatingsyste': Resource(config_templates) {u'short_description': u'List p
rovisioning templates per operating system', u'deprecated': None, u'http_method': u'GET', u'api_url': u'/api/operatingsystem/:operatingsystem_id/config_templates'}
ERROR:root:There is API entry-point which doesn't match expected constraints. Got resource 'media' expected reference 'medi': Resource(operatingsystems) {u'short_description': u'List all operating systems
 for nested medium', u'deprecated': None, u'http_method': u'GET', u'api_url': u'/api/media/:medium_id/operatingsystems'}
WARNING:root:There are unbound methods for 'operatingsystems'. Please see debug logs for more details
WARNING:root:There are unbound methods for 'compute_resources'. Please see debug logs for more details
WARNING:root:There are unbound methods for 'config_templates'. Please see debug logs for more details

**Comment by avi3tal** [comment]
Is it possible that for some reason you automatically removed the last chars? see "medi" and "operatingsyste"

**Comment by avi3tal** [comment]
Here it is:
resources_to_resource():
...
    if not resource:
                resource = resources_name[:-1]
...

Looks like we have two special resources:
media and operatingsystem that doeksn't have 's' suffix.
I guess you should handle them differently.

**Comment by lukas-bednar** [comment]
You are right in case of 'media' resource. but in case of operating system it is different, you can see there is 'operatingsystems' resource as well. this 'operatingsystem' (without 's') coming from 'config_templates' and it looks like a bug to me.

that '/api/operatingsystem/:operatingsystem_id/config_templates' should get map as Foreman.operatingsystems.index_config_templates(os_id) .

Are you still using same json definition (or same instance of Foreman) ? I am not able to reproduce it with the json file you sent me.

**Comment by avi3tal** [comment]
Yes i am using the same instance.
BTW, do you plan to open a bug ticket for Foreman?

File: foreman/client.py [hunk 1]
```diff
@@ -114,24 +125,42 @@ def __new__(meta, name, bases, dct):
                 m = meta.multi_api_reg.search(api['api_url'])
```
Comment (david-caro) [inline foreman/client.py]: s/m/match/

**Comment by david-caro** [comment]
I opened #37 with another version of this fix (it started as a simple modification, now it's more like a refactor).
Can you check if that fixes your issues?

It also fixes some issues like actually requiring the parameters and overwriting functions

**Pull request closed without merging**